wallet: introduce setfeerate (an improved settxfee, in sat/vB)

jonatack opened the pull request:
Continuing the work on issue #19543, this proposes a `setfeerate` RPC in sat/vB and makes `settxfee` a hidden RPC per the plan described in https://github.com/bitcoin/bitcoin/pull/20484#issuecomment-734786305.

jonatack committed:
wallet: introduce setfeerate, an improved settxfee in sat/vB

jonatack committed:
test: update functional tests from settxfee to setfeerate

updates tests that call settxfee to call setfeerate instead;
settxfee functionality still has test coverage in
- wallet_bumpfee.py
- wallet_create_tx.py
- wallet_multiwallet.py

also fixes two fundrawtransaction tests that had off-by-one naming errors
and tightens up one of the two

jonatack committed:
wallet: update settxfee help

jonatack commented:
Thanks to @Xekyo for suggesting this in https://github.com/bitcoin/bitcoin/pull/20305#discussion_r520234818.

jonatack commented:
Idea: since this is per-wallet, return the name of the wallet in the responses.

prayank23 commented:
In the 'Successful responses' and 'Error responses' above shouldn't it be `sat/vB` instead of `sat/B`?

jonatack commented:
@prayank23 thanks for looking. Those will be automatically updated from sat/B to sat/vB by #20305, which updates the ToString() method for CFeeRate.

jonatack committed:
policy: allow CFeeRate::ToString to not append fee rate units

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2347,6 +2347,74 @@ static RPCHelpMan settxfee()
+
+            UniValue obj(UniValue::VOBJ);
+            obj.pushKV("wallet_name", wallet.GetName());
+            obj.pushKV("fee_rate", UniValue(UniValue::VNUM, wallet.m_pay_tx_fee.SatsToString()));
```
IMO we should add a `ValueFromFeeRate`

luke-jr replied:
Maybe rebase this on top of 82b0352144fbe21b60a7d3f9351debec758618a0

jonatack replied:
> Maybe rebase this on top of [8798383](https://github.com/bitcoin/bitcoin/commit/8798383475f4e1db44f734cdd94885e5527761fe)

SGTM

jonatack replied:
> Maybe rebase this on top of [8798383](https://github.com/bitcoin/bitcoin/commit/8798383475f4e1db44f734cdd94885e5527761fe)

Pulled 8798383 in as commit 0e2c3ace86b8aa

luke-jr committed:
core_io: Add ValueFromFeeRate helper

jonatack committed:
wallet, test: upgradewallet follow-ups

per PR 20403 review feedback by MarcoFalke

jonatack committed:
wallet, test: make settxfee a hidden rpc

Make settxfee a hidden rpc to avoid confusion for new users and guide them
to use the setfeerate rpc in sat/vB, while allowing existing scripts based
on settxfee in BTC/kvB to continue using it without breaking.

A hidden RPC means that `bitcoin-cli help settxfee` continues to display
the help documentation while removing it from the results of the general
`bitcoin-cli help`.

Updates the release note.

jonatack committed:
test: add setfeerate functional coverage in wallet_bumpfee.py

jonatack committed:
test: add setfeerate functional coverage in wallet_create_tx.py

jonatack commented:
Rebased and updated per the plan outlined in https://github.com/bitcoin/bitcoin/pull/20484#issuecomment-734786305.

MarcoFalke commented on test/functional/rpc_getblockstats.py:
```diff
@@ -107,7 +107,7 @@ def run_test(self):
 
         for i in range(self.max_stat_pos+1):
-            self.log.info('Checking block %d\n' % (i))
+            self.log.info('Checking block %d' % (i))
```
style-nit: might use f-string?

jonatack replied:
good idea; done

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2350,6 +2350,78 @@ static RPCHelpMan settxfee()
+
+            LOCK(wallet.cs_wallet);
+
+            CFeeRate amount{AmountFromValue(request.params[0]), COIN};
```
The second argument of the cfeerate constructor is size_t (number of bytes), not CAmount (number of satoshis). Instead of adding more calls to the confusing constructor, I'd prefer to add a new constructor that takes in a univalue feerate as sat/b. I think this has also been provided as feedback when this code template was initially introduced.

jonatack replied:
Good idea, done in 9c479bfc293fec906 and proposed a refactoring in 392495bc4d9648fe0c28d that can be dropped

jonatack replied:
Hm, bringing the RPC utilities and UniValue into CFeeRate seems like a layer violation. Maybe a sat/vB CFeeRate ctor that takes a CAmount instead of UniValue.

jonatack replied:
Updated the new CFeeRate ctor to take `(CAmount fee_rate, FeeEstimateMode mode)` and added unit tests.

maflcko reviewed (approved):
Concept ACK on adding the feature for users to set -paytxfee via rpc at runtime in sat/b

jonatack committed:
policy: add CFeeRate::FromSatB/FromBtcKb constructors

jonatack commented:
Oops, adding the CFeeRate UniValue ctor caused a circular dependency.

jonatack committed:
test: add ValueFromFeeRate/CFeeRate unit tests

jonatack commented:
Added more unit test coverage per `git diff 14c5e50 2176a3`. This should hopefully be ready now.

MarcoFalke commented on src/policy/feerate.cpp:
```diff
@@ -42,3 +42,7 @@ std::string CFeeRate::ToString(const FeeEstimateMode& fee_estimate_mode) const
     }
 }
+
+std::string CFeeRate::SatsToString() const {
```
style nit d6643ef609bf442fe38dd8a1335dd7c580f1fdd7:

could clang-format?

jonatack replied:
no longer adding this `SatsToString()` function in favor of making the existing `ToString()` more flexible

MarcoFalke commented on src/test/amount_tests.cpp:
```diff
@@ -86,6 +90,31 @@ BOOST_AUTO_TEST_CASE(GetFeeTest)
+    BOOST_CHECK(CFeeRate(CAmount{123}, FeeEstimateMode::UNSET) == CFeeRate{0});
+    BOOST_CHECK(CFeeRate(CAmount{123}, std::numeric_limits<size_t>::max()) == CFeeRate{0});
+    BOOST_CHECK(CFeeRate(CAmount{123}, 0) == CFeeRate{0});
+    BOOST_CHECK(CFeeRate(CAmount{123}, -1) == CFeeRate{0});
```
style nit in 4cccf2d88377a9f813ab6d63889cbb5136c31aad

Might be good to make this impossible to compile instead of a runtime-just-return-0. Can be achieved via `constexpr if` (and some template magic) or by having a `CFeeRate::FromSatB(amount)` and `FromBtcKb(amount)` helper.

jonatack replied:
Oh oops hm, these lines 113-115 aren't testing the new constructor, but the `CFeeRate(CAmount, size_t)` one, so it was mistake to add them here. Will move them into the tests for that constructor and see if anything needs to be done there; normally the ctor should just set `nSatoshisPerK` to 0 in that case.

MarcoFalke replied:
(my comment should be on the `BOOST_CHECK(CFeeRate(CAmount{123}, FeeEstimateMode::UNSET) == CFeeRate{0})` line)

jonatack replied:
Good idea, much better...switching to named constructors.

jonatack replied:
done, and a leaner implementation, thanks

MarcoFalke commented on src/core_io.h:
```diff
@@ -40,6 +41,7 @@ int ParseSighashString(const UniValue& sighash);
 
 // core_write.cpp
 UniValue ValueFromAmount(const CAmount& amount);
+UniValue ValueFromFeeRate(const CFeeRate& fee_rate);
```
in commit 3c92185b95764f5c3536637e5923316cd9c2d8b0:

Might be good if the method name included the unit (sat/b) to avoid calling this in a btc/kb context

jonatack replied:
Updated to allow passing in the fee mode to work with both sat/vB and BTC/kvB

maflcko reviewed (commented):
left some comments

MarcoFalke commented on src/policy/feerate.cpp:
```diff
@@ -7,35 +7,31 @@
 
```
why are the refactors in 7289e559eb720d972e3e31c0ca81dd8091775b07 needed? They break asan, it seems.

jonatack replied:
The CI was green with the refactors yesterday; the change was the added unit tests. It may be better to propose the refactors in their own PR. I'll drop the refactoring commit to test what the CI says.

jonatack committed:
wallet: update rpcwallet to CFeeRate named constructors

jonatack commented:
Thanks for the review @MarcoFalke. Dropped the policy/feerate refactoring commit and the FeeModeToBytes commit, changed the new CFeeRate delegating constructor to named constructors, dropped SatsToString in favor of making the existing ToString more flexible, and made ValueFromFeeRate able to work with both sat/b and btc/kb.

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2350,6 +2350,79 @@ static RPCHelpMan settxfee()
+        "setfeerate",
+        "\nSet the transaction fee rate in " + CURRENCY_ATOM + "/vB for this wallet.\n"
+        "Overrides the global -paytxfee configuration option.\n"
+        "Can be deactivated by passing 0 as the fee rate, in which case automatic fee selection will be used by default.\n",
```
c6e9bfd1bd9d6a65673fbe1c278a7c7ada35273d

Worth noting that the new settings is not persisted?

jonatack replied:
Thanks, added to settxfee and setfeerate helps

promag commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2350,6 +2350,79 @@ static RPCHelpMan settxfee()
+            CHECK_NONFATAL(result.empty() != error.empty());
+
+            UniValue obj{UniValue::VOBJ};
+            obj.pushKV("wallet_name", wallet.GetName());
```
c6e9bfd1bd9d6a65673fbe1c278a7c7ada35273d

Why return wallet name? What about returning previous effective value?

MarcoFalke replied:
why would the user care about the previous value? If they did, they could call getwalletinfo

promag replied:
Right, the same goes for the wallet name.

jonatack replied:
I'd prefer to keep wallet name.

jonasschnelli replied:
No strong opinion about the wallet name in the response. I think it does not hurt and in asynchronous environments it might simplify the client code.

MarcoFalke replied:
I think we should establish a guideline first on what input parameters to return. I don't think we've done this in the past and this could lead to user confusion: "What am I supposed to do with the return parameter that is equal to the input parameter? Assert that they are the same? Ignore?"

If this is something useful to return for every wallet RPC, then all wallet RPCs should be updated.

jonatack replied:
I see it as good for rpcs that make per-wallet changes, like here and upgradewallet as a recent example (see upgradewallet help). It would also be good in rpc/cli options that return info on a per-wallet basis, like getwalletinfo but also -rpcwallet= -getinfo, for example.

promag reviewed (commented):
Code review ACK 73a0f0276392e4432774807f2e4b8a3f4dfc69ab, needs to squash some commits. Also add release notes?

jonatack commented:
Updated with two fixes:

- Checks for and raises if the passed value is non-representable by CFeeRate (lesson learned from #20546), added test coverage

- Fixes the error message if the current value is unset (indicates a current value of "0 (unset)" instead of "0.000 sat/vB"), updated the test coverage

jonatack commented:
Thanks @promag and @MarcoFalke for the review feedback. Updates in latest push:

- Check for and raise "Invalid amount" if the passed value is non-representable by CFeeRate (lesson learned from #20546)

- Fix the error messages if the current value is unset, to indicate a current value of "0 (unset)" instead of "0.000 sat/vB"

- Make sure the tests check passing amounts with a string as well as a number, and update the coverage for the first two items

- Update the settxfee and setfeerate helps with @promag's suggestion

jonatack commented:
I'll re-push with a release note in f4195d2a94b

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2350,6 +2352,81 @@ static RPCHelpMan settxfee()
+            std::string result, error;
+
+            if (amount == zero) {
+                if (request.params[0].get_real() != 0) throw JSONRPCError(RPC_TYPE_ERROR, "Invalid amount");
```
This should be thrown in the amountfromvalue function. Doing it here with the zero-not-zero hack smells a bit like a layer violation

jonatack replied:
> This should be thrown in the amountfromvalue function.

thanks -- was thinking the same

jonatack replied:
After #20546 this will go away:
```diff
-            const CFeeRate amount{CFeeRate::FromSatB(AmountFromValue(request.params[0]))};
+            const CFeeRate amount{FeeRateFromValueInSatB(request.params[0])};
@@ -2399,7 +2399,6 @@ static RPCHelpMan setfeerate()
 
             if (amount == zero) {
-                if (request.params[0].get_real() != 0) throw JSONRPCError(RPC_TYPE_ERROR, "Invalid amount");
                 wallet.m_pay_tx_fee = amount;
```

jonatack replied:
This PR has been gated since a month on the changes in #20546 to handle this properly, but that PR didn't go anywhere, so I'll drop this check for now.

jonatack replied:
Hm, realized I need to leave this check or else will have to drop a bunch of tests, which would be a shame and a regression. A lot of work went into this and it was waiting for #20546. I guess  leaving it as-is for now.

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2311,10 +2311,12 @@ static RPCHelpMan listlockunspent()
     return RPCHelpMan{"settxfee",
-                "\nSet the transaction fee per kB for this wallet. Overrides the global -paytxfee command line parameter.\n"
-                "Can be deactivated by passing 0 as the fee. In that case automatic fee selection will be used by default.\n",
+                "\n(Superseded by the setfeerate RPC in sat/vB, which it is recommended to use instead.)\n"
```
Wouldn't `deprecated` instead of `superseded` be more consistent with previous RPC API "changes"?

jonatack replied:
Thanks, will do on next update.

jonatack replied:
done

jonasschnelli reviewed (approved):
Concept ACK (partially code reviewed).

jonatack commented:
> Code review ACK [73a0f02](https://github.com/bitcoin/bitcoin/commit/73a0f0276392e4432774807f2e4b8a3f4dfc69ab), needs to squash some commits. Also add release notes?

Rebased, updated setfeerate help with @jonasschnelli's suggestion, added release notes per @promag's suggestion. Sorry for the delay in updating, this was waiting since a month on the improvements and fixes in #20546 but I've had to move forward without them. I think this is done. Please re-ACK if you can.

S3RK commented:
I start reviewing this one.

S3RK commented on src/test/amount_tests.cpp:
```diff
@@ -86,6 +90,26 @@ BOOST_AUTO_TEST_CASE(GetFeeTest)
+    BOOST_CHECK(CFeeRate::FromSatB(CAmount(100000)) == CFeeRate(1));
+    BOOST_CHECK(CFeeRate::FromSatB(CAmount(100001)) == CFeeRate(1));
+    BOOST_CHECK(CFeeRate::FromSatB(CAmount(2690000)) == CFeeRate(26));
+    BOOST_CHECK(CFeeRate::FromSatB(CAmount(123456789)) == CFeeRate(1234));
```
what happens if we pass an absurdly big number?

S3RK reviewed (commented):
I did a high level pass, didn't look at functional tests at all yet. Overall seems good, but I need to give it more thorough look. Please check my comments.

jonatack commented on src/policy/feerate.cpp:
```diff
@@ -35,10 +35,21 @@ CAmount CFeeRate::GetFee(size_t nBytes_) const
-    default:                      return strprintf("%d.%08d %s/kvB", nSatoshisPerK / COIN, nSatoshisPerK % COIN, CURRENCY_UNIT);
+    if (with_units) {
+        switch (mode) {
+        case FeeEstimateMode::SAT_VB:
```
IIRC that was done in https://github.com/bitcoin/bitcoin/pull/11413, but good point. I'll look at fixing that.

jonatack replied:
Ah, I have a pull request that does that but have not opened it yet to not have too many PRs open.

jonatack replied:
Ah, just remembered that I wrote a patch in November 2020 that does that but I have not opened it yet to not have too many PRs open.

jonatack commented:
Thank you @S3RK for your feedback, much appreciated. I'll reply soon.

S3RK commented on src/wallet/rpcwallet.cpp:
```diff
@@ -216,7 +216,7 @@ static void SetFeeEstimateMode(const CWallet& wallet, CCoinControl& cc, const Un
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Cannot specify both estimate_mode and fee_rate");
         }
-        cc.m_feerate = CFeeRate(AmountFromValue(fee_rate), COIN);
+        cc.m_feerate = CFeeRate::FromSatB(AmountFromValue(fee_rate));
```
I love the idea of named constructors. Should we strive for consistency and use them in more places?

S3RK commented on test/functional/wallet_create_tx.py:
```diff
@@ -77,6 +79,51 @@ def test_tx_size_too_large(self):
+        def test_response(*, requested=0, expected=0, error=None, msg):
+            assert_equal(node.setfeerate(requested), {"wallet_name": "test_wallet", "fee_rate": expected, ("error" if error else "result"): msg})
+
+        # Test setfeerate with 10.0001 (CFeeRate rounding), "10.001" and "4" sat/vB
```
Again I believe those check would be better places in a separate `setfeerate` test

S3RK reviewed (commented):
I finished the review. A bunch of small questions/comments, but nothing serious. Happy to discuss and do further reviews.

luke-jr commented on test/functional/wallet_bumpfee.py:
```diff
@@ -315,6 +317,42 @@ def test_dust_to_fee(self, rbf_node, dest_address):
+    test_response(requested=fee_rate, expected=fee_rate, msg="Fee rate for transactions with this wallet successfully set to 25.000 sat/vB")
+    bumped_tx = rbf_node.bumpfee(rbfid)
+    actual_feerate = bumped_tx["fee"] * COIN / rbf_node.getrawtransaction(bumped_tx["txid"], True)["vsize"]
+    assert_greater_than(Decimal("0.01"), abs(fee_rate - actual_feerate))
```
Even 0.01 is too small for this. Signatures can be 71-73 bytes long, but even 72 bytes throws this off.

```suggestion
    bumped_txdetails = rbf_node.getrawtransaction(bumped_tx["txid"], True)
    allow_for_bytes_offset = len(bumped_txdetails['vout']) * 2  # potentially up to 2 bytes per output
    actual_fee = bumped_tx["fee"] * COIN
    assert_approx(actual_fee, fee_rate * bumped_txdetails['vsize'], fee_rate * allow_for_bytes_offset)
```

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -216,7 +216,7 @@ static void SetFeeEstimateMode(const CWallet& wallet, CCoinControl& cc, const Un
             throw JSONRPCError(RPC_INVALID_PARAMETER, "Cannot specify both estimate_mode and fee_rate");
         }
-        cc.m_feerate = CFeeRate(AmountFromValue(fee_rate), COIN);
+        cc.m_feerate = CFeeRate::FromSatB(AmountFromValue(fee_rate));
```
In commit "wallet: update rpcwallet to CFeeRate named constructors" (c72083c2d56237f8fbd3a82e3ff395f0cdb8fd57)

Wow wow. I see that this is doing the right thing, but these constructor names seem absurdly misleading! The `AmountFromValue` function converts a floating point number to a fixed-point representation of that number multiplied by `10e8`. So even though the `fee_rate` univalue argument has units of sat/B, the `AmountFromValue` integer return value is no longer in the same units, because it has been multiplied by 100000000. So `FromSatB` should really be called `CFeeRate::FromSatB_Fixed_Point_e8` or something like that because it is taking sat/B values with a very specific, atypical number representation.

An analogous thing is happening with `FromBtcKb` below. It doesn't really take a BTC/Kb argument. That's basically false advertising. What it really takes is a BTC/Kb*10e8 argument which is the same as a sat/Kb argument, so this constructor would be more accurately called `CFeeRate::FromSatKb`

I guess I'm having a negative reaction to these constructor names, but I do think this could be cleaned up pretty easily. I think it would be better if `CFeeRate` class had no involvement with univalue parsing or fixed point representations, and if it didn't (mis)use the CAmount type as a way to represent rates when `CAmount` is used almost everywhere else only to represent quantities.

My suggestion to be less confusing / misleading would be to drop the two new `CFeeRate` constructors, and just add standalone RPC util functions:

```c+++
CFeeRate FeeRateFromSatB(const UniValue& value);
CFeeRate FeeRateFromBtcKb(const UniValue& value);
```

Then call these to simplify `CFeeRate::FromSatB(AmountFromValue(feerate))` as `FeeRateFromSatB(feerate)` here and in `setfeerate` and to simplify `CFeeRate::FromBtcKb(AmountFromValue(feerate))` as `FeeRateFromBtcKb(feerate)` in `settxfee`

MarcoFalke replied:
> Wow wow. I see that this is doing the right thing, but these constructor names seem absurdly misleading!

The current code in master isn't exactly clear right now either:

```cpp
    /** Constructor for a fee rate in satoshis per kvB (sat/kvB).
     *
     *  Passing a num_bytes value of COIN (1e8) returns a fee rate in satoshis per vB (sat/vB),
     *  e.g. (nFeePaid * 1e8 / 1e3) == (nFeePaid / 1e5),
     *  where 1e5 is the ratio to convert from BTC/kvB to sat/vB.
     */
    CFeeRate(const CAmount& nFeePaid, uint32_t num_bytes);
```

I think it would be good to figure out how to clear this up before introducing further features that depend on this.

MarcoFalke replied:
Original discussion for reference: https://github.com/bitcoin/bitcoin/pull/20305#discussion_r522371491

jonatack replied:
> > Wow wow. I see that this is doing the right thing, but these constructor names seem absurdly misleading!
> 
> The current code in master isn't exactly clear right now either:
> 
> ```c++
>     /** Constructor for a fee rate in satoshis per kvB (sat/kvB).
>      *
>      *  Passing a num_bytes value of COIN (1e8) returns a fee rate in satoshis per vB (sat/vB),
>      *  e.g. (nFeePaid * 1e8 / 1e3) == (nFeePaid / 1e5),
>      *  where 1e5 is the ratio to convert from BTC/kvB to sat/vB.
>      */
>     CFeeRate(const CAmount& nFeePaid, uint32_t num_bytes);
> ```
> 
> I think it would be good to figure out how to clear this up before introducing further features that depend on this.

Ok, seems best to close this then. It's almost a year old and needs to be done differently.  I think using sat/vB is still a popular request from users and plan to try again a bit later.

ryanofsky reviewed (commented):
Started review (will update list below with progress). I got a little sidetracked on the cfeerate constructor names, but overall this seems like a very nice, well thought out PR

- [X] c4b288d5d9ac0035c1f2d935aa7d80d86d76a010 policy: add CFeeRate::FromSatB/FromBtcKb constructors (1/12)
- [X] c72083c2d56237f8fbd3a82e3ff395f0cdb8fd57 wallet: update rpcwallet to CFeeRate named constructors (2/12)
- [ ] 5ed3ca1fafcfe93c35737c26f6e826ed451081fa policy: allow CFeeRate::ToString to not append fee rate units (3/12)
- [ ] aee3571392d095412ac6bda132ec6b1db6c42774 core_io: Add ValueFromFeeRate helper (4/12)
- [ ] 0479268c0830caea8cb6ff344531d85417b99d02 test: add ValueFromFeeRate/CFeeRate unit tests (5/12)
- [ ] 8e863e3d3ce457c9ca26a12e8fd6beac2f50aa7d wallet: introduce setfeerate, an improved settxfee in sat/vB (6/12)
- [ ] 529bfc16ffb35c5356e27f61d59395fae6707bcd test: add setfeerate functional coverage in wallet_create_tx.py (7/12)
- [ ] c907f158a6bf3cad782d4441e02abcbda210265b test: add setfeerate functional coverage in wallet_bumpfee.py (8/12)
- [ ] d87f0f3a923dca2ce8d99aa700b0beda8e42d3ec test: update functional tests from settxfee to setfeerate (9/12)
- [ ] c594a002845e7003319252961fc35101fcb1496e wallet: update settxfee help (10/12)
- [ ] 3725a3f180fa8f6edeee2fdca3e09b6022c5eb21 wallet, test: make settxfee a hidden rpc (11/12)
- [ ] 1002e2d0d7f10221a3542391a429978cdcf0d426 wallet, test: upgradewallet follow-ups (12/12)

jonatack commented:
Thanks! Huge apologies, since the merge of https://github.com/bitcoin/bitcoin/pull/21786 that resolved some of these things differently (at a lower level), the approach of the initial refactoring commits is obsolete and being changed. I need to push an update and have been intending to do so for some time. 🙏
